### ISSUE TITLE
Latest version of chrome causes capybara specs to fail with Net::ReadTimeout

### ISSUE THREAD

**Issue by superchris** [teamcapybara/capybara#2181]
New version of chrome (74) hangs and eventually dies with Net::ReadTimeout. Downgrading to 73 fixes the issue.

## Meta
Capybara Version: 3.4.0

Driver Information (and browser if relevant):
Chrome 7.

## Actual Behavior
Specs work.

## Steps to reproduce
Run any javascript spec

**Comment by iwz** [comment]
Also experiencing this on Capybara 2.18.0 (Rails 3.2). We're using the [`webdrivers 3.8.0` gem](https://github.com/titusfortner/webdrivers) to get the correct version of `chromedriver` for our currently installed version of Chrome (74). When running capybara `:js` tests, we get many `Net::ReadTimeout` errors and the tests fail.

On macOS, we are able to work around this by following these steps:
- Remove chrome
- download v 72 here (https://google-chrome.en.uptodown.com/mac/old) (cannot find a version of Chrome 73 anywhere...)
- run `sudo chmod -R 000 ~/Library/Google` to prevent chrome from having permissions to execute the updater script, and it can’t update itself any more (unless you run `sudo chmod -R 755 ~/Library/Google` and allow it to).

**Comment by elliterate** [comment]
Possibly the same as https://bugs.chromium.org/p/chromedriver/issues/detail?id=2878?

**Comment by bgwilson87** [comment]
Sample failure, using rails 5.0.7.2:
```
Net::ReadTimeout:
  Net::ReadTimeout
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/http/default.rb:129:in `response_for'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/http/default.rb:82:in `request'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/http/common.rb:64:in `call'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/bridge.rb:167:in `execute'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/oss/bridge.rb:587:in `execute'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/remote/oss/bridge.rb:154:in `window_handles'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/driver.rb:187:in `window_handles'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/capybara-3.18.0/lib/capybara/selenium/driver.rb:192:in `window_handles'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/capybara-3.18.0/lib/capybara/selenium/driver_specializations/chrome_driver.rb:34:in `reset!'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/capybara-3.18.0/lib/capybara/session.rb:128:in `reset!'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/capybara-3.18.0/lib/capybara.rb:315:in `block in reset_sessions!'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/capybara-3.18.0/lib/capybara.rb:315:in `reverse_each'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/capybara-3.18.0/lib/capybara.rb:315:in `reset_sessions!'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/capybara-3.18.0/lib/capybara/rspec.rb:18:in `block (2 levels) in <top (required)>'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:123:in `block in run'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `loop'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `run'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/rspec-retry-0.6.1/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
# ./spec/spec_helper.rb:31:in `block (2 levels) in <top (required)>'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:123:in `block in run'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `loop'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `run'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/rspec-retry-0.6.1/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
# /Users/my_user/.rvm/gems/ruby-2.5.5/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:37:in `block (2 levels) in setup'
```

**Comment by twalpole** [comment]
If it works fine with Chrome 73 but 74 times out that would tend to imply an issue with Chrome of chromedriver. Without a way to replicate the issue I can't even really attempt to diagnose anything.  Capybaras test suite runs against Chrome 74 without issue - https://travis-ci.org/teamcapybara/capybara/jobs/525827115 .   If someone can provide a way to replicate I can look further ...

**Comment by bgwilson87** [comment]
This is the config that I'm using, if that helps:

```
Capybara.register_driver :chrome do |app|
  Capybara::Selenium::Driver.new(app, browser: :chrome)
end

Capybara.register_driver :headless_chrome do |app|
  capabilities = Selenium::WebDriver::Remote::Capabilities.chrome(
    chromeOptions: { args: %w(headless disable-gpu window-size=1280,2000 no-sandbox) }
  )

  Capybara::Selenium::Driver.new app,
    browser: :chrome,
    desired_capabilities: capabilities
end

Capybara.javascript_driver = :headless_chrome
```

I have noticed that when I remove headless from chromeOptions, it works.  It very well may be a chrome / chromedriver issue.

**Comment by iwz** [comment]
Here's another sample config:

```ruby
require 'webdrivers'
require 'selenium-webdriver'

Capybara.register_driver :chrome do |app|
  Capybara::Selenium::Driver.new(app, browser: :chrome)
end

Capybara.register_driver :headless_chrome do |app|
  capabilities = Selenium::WebDriver::Remote::Capabilities.chrome(
    chromeOptions: { args: %w(headless window-size=1024,768) },
    loggingPrefs: { browser: 'ALL' }
  )

  Capybara::Selenium::Driver.new app,
    browser: :chrome,
    desired_capabilities: capabilities
end

Capybara.default_max_wait_time = 5
Capybara.javascript_driver = :headless_chrome # :chrome simulates behavior in browser
Capybara.server_port = 31337 + ENV['TEST_ENV_NUMBER'].to_i
Capybara.configure do |config|
  config.always_include_port = true
end
```

And the basic test case is (with Chrome 74, and `webdrivers`):

```ruby
require "spec_helper.rb"

visit "/"
```

**Comment by twalpole** [comment]
@bgwilson87 @iwz Can you try adding the `disable-features=VizDisplayCompositor` chrome option and see if that makes any difference

**Comment by iwz** [comment]
> @bgwilson87 @iwz Can you try adding the `disable-features=VizDisplayCompositor` chrome option and see if that makes any difference

@twalpole this does not make a difference for me.

**Comment by twalpole** [comment]
@iwz Ok then I think it's a chromedriver issue where there's nothing Capybara can really do about it.  It doesn't affect the Capybara tests on travis or locally for me on MacOS 10.14.4 so it's possible it's a platform specific issue in chromedriver. Does it occur in the same code location for you? (in session reset when Capybara is asking chromedriver for a list of the open windows)

**Comment by bgwilson87** [comment]
I have opened an issue in chromedriver here https://bugs.chromium.org/p/chromedriver/issues/detail?id=2885

**Comment by ArthurGIT2017** [comment]
The following Ruby code didn't reproduce this issue:

Selenium::WebDriver::Chrome.driver_path = File.join("<path_to_v74_chromedriver>/chromedriver")
options = Selenium::WebDriver::Chrome::Options.new(:binary=>"<path_to_v74_chrome>/chrome")
options.add_argument('--headless')
driver=Selenium::WebDriver.for :chrome, options: options
driver.get 'http://google.com/'

**Comment by iwz** [comment]
There is a workaround: disable `--headless`. With headless turned off, the Capybara and chromedriver work fine with Chrome 74.

**Comment by RaipheKelsch** [comment]
We are also experiencing this issue with all versions 3.16 and above let me know if you need further information.

**Comment by twalpole** [comment]
@RaipheKelsch The further information we need is a self contained way to reproduce this issue.  As stated above the Capybara test suite runs fine against Chrome/chromedriver 74 on TravisCI (in both headless and headed configurations).  

Since you've mentioned Capybara 3.16. It made a change over to using a persistent connection which could theoretically cause issues if you are running against remote connections that prevent persistent connections for some reason, however I believe that would be a different issue than the one being discussed here.

**Comment by CasJam** [comment]
it seems the Chrome 74 upgrade broke my tests as well.  I'm seeing this error:

```
Selenium::WebDriver::Error::SessionNotCreatedError:
        session not created: Chrome version must be between 70 and 73
          (Driver info: chromedriver=73.0.3683.20 (8e2b610813e167eee3619ac4ce6e42e3ec622017),platform=Mac OS X 10.13.6 x86_64)
```
Chrome auto-updated to Version 74.0.3729.131

I tried downgrading to v72 using the link from @iwz's post above, installed it, my tests passed, but I was unable to launch Chrome on my mac with that version (it showed some error telling me "Chrome is damaged, move it to trash".

So now I'm back to running the latest Chrome version, and all my JS tests fail...

Waiting for someone to provide instructions on what to do!

**Comment by twalpole** [comment]
@casjam The chromedriver version you’re using needs to match the chrome version . If you’re using Chrome 74.x you need chromedriver 74.x to go with it

**Comment by greena13** [comment]
I've found a work-around that is sufficient for my needs; it may help others:

```
# @TODO Remove this Monkey Patch after issue is resolved with running Chrome 74 in
#       headless mode: https://github.com/teamcapybara/capybara/issues/2181
module Selenium
  module WebDriver
    class Options
      # capybara/rspec installs a RSpec callback that runs after each test and resets
      # the session - part of which is deleting all cookies. However the call to Chrome
      # Webdriver to delete all cookies in Chrome 74 hangs when run in headless mode 
      # (the reasons for which are still unknown).
      #
      # Fortunately, the call to set a cookie is still functioning and we can rely
      # on expired cookies being cleared by Chrome, so we iterate over all current
      # cookies and set their expiry date to some time in the past - effectively
      # deleting them.
      def delete_all_cookies
        all_cookies.each do |cookie|
          add_cookie(name: cookie[:name], value: '', expires: Time.now - 1.second)
        end
      end
    end
  end
end
```

The issue appears to happen whenever a page has some sort of external dependency (an asynchronous script, or an iframe with content from another domain, for example). I am not sure, but suspect it may be whenever an cookie for an external domain is set. 

I am, however, reasonably sure this is in fact an issue with Chrome or Chrome Driver, and not Capybara.

If anyone wishes to debug the underlying cause or develop a better workaround the one above, you can configure the chrome driver to log out to file, like so:

```
Capybara.register_driver :headless_chrome do |app|
  capabilities = Selenium::WebDriver::Remote::Capabilities.chrome(
    chromeOptions: { args: %w(headless disable-gpu) }
  )

  Capybara::Selenium::Driver.new app,
                                 browser: :chrome,
                                 desired_capabilities: capabilities,
                                 driver_opts: {
                                   verbose: true,
                                   log_path: 'path/to/app/tmp/chromedriver.log'
                                 }
end
```

**Comment by greena13** [comment]
@CasJam, you may also find it useful to temporarily enable the webdriver gem's logging to confirm what version of the Chrome webdriver you are using (it will log as much to the console the next time you run your test suite):

```
Webdrivers.logger.level = :DEBUG
```

**Comment by CasJam** [comment]
thanks @twalpole @greena13 I updated chromedriver to `74.0.3729.6` and tests are working again.

**Comment by pixiitech** [comment]
@greena13 I'm trying to implement your patch, but I don't know where to put this file... I tried ```config/initializers/selenium.rb``` but this did not work.

**Comment by greena13** [comment]
@pixiitech, I put it in a file in the /spec/support/ directory and then required that file in /spec/spec_helper.rb.

I don't have the code in front of me at the moment, but you may need to play around with the order that you require it relative to where you require capybara.

**Comment by jackcasey** [comment]
@greena13 Thanks for that patch! I was about to eat my keyboard. It works for me.

**Comment by twalpole** [comment]
@greena13 Are you running against a remote instance of Chrome, or a local one?

**Comment by jeppeliisberg** [comment]
@greena13 your patch didn't work for me unfortunately :-( I'm still stuck with timeouts in my local development. 

However, tests run fine on Heroku CI using https://github.com/heroku/heroku-buildpack-google-chrome and https://github.com/heroku/heroku-buildpack-chromedriver which in turn download the latest chrome and chromdriver (logs says `-----> Downloading chromedriver v74.0.3729.6...`).

I'm looking into it further as time allows and will post any findings here...

**Comment by jeppeliisberg** [comment]
Quick update - after having tried with chrome beta (v75) I reinstalled stable (v74), downloaded chromedriver v74.0.3729.6 and moved it to `/usr/local/bin/` and now my test suite runs again WITHOUT the patch from @greena13.

I can't explain what happened though.

My configuration:
```
chrome_bin = ENV.fetch("GOOGLE_CHROME_SHIM", nil)
chrome_opts = chrome_bin ? { "chromeOptions" => { "binary" => chrome_bin } } : { "chromeOptions" => { "args" => %w[headless no-sandbox] } }

Capybara.register_driver :chrome do |app|
  Capybara::Selenium::Driver.new(
    app,
    browser: :chrome,
    desired_capabilities: Selenium::WebDriver::Remote::Capabilities.chrome(chrome_opts)
  )
end
Capybara.javascript_driver = :chrome
```

The `GOOGLE_CHROME_SHIM` is for Heroku CI (see earlier post).

**Comment by ar31an** [comment]
I have had the same issue with capybara specs on mac. Adding this argument solved the issue:

Reference: https://bugs.chromium.org/p/chromedriver/issues/detail?id=2897

**Comment by nruth** [comment]
I've been getting weird test failures since Chrome 74 installed itself (initially the Net::ReadTimeout, then after moving to webdrivers it just broke content based browser synchronisation or introduced some session cleanup race condition vs running without headless) and would like to suggest another work-around for people who don't have a Docker/VM testing setup. It unfortunately does nothing to help find or fix the problem, but might save someone else some time. 

# Chromium: A Downgrading Solution

Old Chrome versions aren't officially made available, and getting it from a third party is dangerous. We've got another option though: Chromium, which does make old builds available. And they don't auto-update.

This is suitable for folk who want their tests to run and don't care which browser/version they run in, but don't want to spend time re-stabilising their tests in headless Firefox. This assumes use of the `webdrivers` gem. I've not tried it with chromedriver-helper.

I've used this approach on my Mac to get my tests running locally again without having to remove the headless flag. 

1. Download a Chromium build corresponding with Chrome 73, as per https://www.chromium.org/getting-involved/download-chromium – I went with https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Mac/625897/
2. Move the app file to your home directory or project directory or wherever you are comfortable with
3. Reconfigure `webdrivers` chrome path to point at this version of chrome
4. Run tests

My specific setup code is this before registering my capybara drivers:

```
if (chrome_path = ENV["WEBDRIVER_CHROME_PATH"])
  Selenium::WebDriver::Chrome.path = chrome_path
end
```

And a dotenv file `.env.test.local` to set the path:

```
WEBDRIVER_CHROME_PATH=/Volumes/dev/bin/Chromium.app/Contents/MacOS/Chromium
```

`webdrivers` managed to switch driver version from 74 to `ChromeDriver 73.0.3683.68` and my tests all pass again.

I like this .env approach because switching back to Chrome is as simple as removing that line from the dotenv file, which is gitignored and will only affect my installation. No need to commit/share anything. It does mean each developer will need to do it independently though so YMMV.


Security Disclaimer: using an old unofficial browser build is introducing a security flaw in your system. Consider firewalling/sandboxing it if you can. In my case I'm restricting its network access to where I expect my tests to go with LittleSnitch.

**Comment by melostbr** [comment]
In a mix with @ar31an and @nruth solutions, I tried to downgrade the Chrome and chromedriver until there's a fix for this. My configuration is now like this:
```
# rails_helper.rb

require "selenium/webdriver"

Capybara.register_driver :headless_chrome do |app|
  capabilities = Selenium::WebDriver::Remote::Capabilities.chrome(
    chromeOptions: {
      binary: "<path_to>/Chromium.app/Contents/MacOS/Chromium",
      args: %w[
        headless disable-gpu no-sandbox
        --window-size=1980,1080 --enable-features=NetworkService,NetworkServiceInProcess
      ]
    }
  )

  Capybara::Selenium::Driver.new app, browser: :chrome, desired_capabilities: capabilities
end
```

When I run my tests, I get this error:
```
Selenium::WebDriver::Error::SessionNotCreatedError:
            session not created: Chrome version must be between 70 and 73
              (Driver info: chromedriver=73.0.3683.68 (47787ec04b6e38e22703e856e101e840b65afe72),platform=Mac OS X 10.14.4 x86_64)
```

I still can't make this work, does someone have an idea where am I doing something wrong?

**Comment by maxshelley** [comment]
For those using Heroku, I just wanted to combine what @ar31an and @jeppeliisberg mention above in case you need another example. Adding the additional argument to the code that Heroku provide in their [Google Chrome buildpack documentation](https://elements.heroku.com/buildpacks/heroku/heroku-buildpack-google-chrome) resolved my issue locally without needing to install Chrome Beta and it continued to work on Heroku .

Here's the configuration that worked for me:

``` ruby
require "selenium/webdriver"

chrome_bin = ENV.fetch('GOOGLE_CHROME_SHIM', nil)

chrome_opts = chrome_bin ? { "chromeOptions" => { "binary" => chrome_bin } } : { chromeOptions: { args: ['headless', 'disable-gpu', '--enable-features=NetworkService,NetworkServiceInProcess'] } }

Capybara.register_driver :chrome do |app|
  Capybara::Selenium::Driver.new(
     app,
     browser: :chrome,
     desired_capabilities: Selenium::WebDriver::Remote::Capabilities.chrome(chrome_opts)
  )
end

Capybara.javascript_driver = :chrome

```

Hope that helps someone!

**Comment by twalpole** [comment]
Has anyone that experienced this issue with Chrome 74 tried Chrome 75? If so is the issue solved on 75 without needing the NetworkService and NetworkServiceInProcess options?

**Comment by Confusion** [comment]
I've been experiencing this issue, but with chromium and chromedriver 73, in both headless and non-headless mode (where the chromium and chromedriver run in a Docker container). Locally I observe *many* sockets (running in the thousands after a while) in TIME_WAIT, reminiscent of earlier problems with correctly closing connections or lack of keepalive.

Capybara 3.18.0, selenium-webdriver 3.142.0

**Comment by twalpole** [comment]
@Confusion If you're using Chrome/chromedriver 73 it's a different issue than this thread is talking about (which is for 74).  Since you're running in a container it's probably you need to specify a combination of chrome options such as `--no-sandbox`, `--disable-dev-shm-usage` and  `--disable-features=VizDisplayCompositor`.

**Comment by catmando** [comment]
per @ar31an 's comment - this worked, and I also found this issue chromedriver issue https://groups.google.com/forum/#!msg/chromedriver-users/ktp-s_0M5NM/9VQO3CKFAAAJ

**Comment by twalpole** [comment]
Closing this since it is a chrome/chromedriver issue, with an available fix, and not a Capybara issue

**Comment by uLucasFraga** [comment]
> I have had the same issue with capybara specs on mac. Adding this argument solved the issue:
> 
> `--enable-features=NetworkService,NetworkServiceInProcess`
> 
> Reference: https://bugs.chromium.org/p/chromedriver/issues/detail?id=2897#c4
> 
> My configuration:
> 
> ```
>   capabilities = Selenium::WebDriver::Remote::Capabilities.chrome(
>     chromeOptions: {
>       args: %w[
>         headless disable-gpu no-sandbox
>         --window-size=1980,1080 --enable-features=NetworkService,NetworkServiceInProcess
>       ]
>     }
>   )
> 
>   Capybara::Selenium::Driver.new app, browser: :chrome, desired_capabilities: capabilities
> ```

That was amazing!
It worked for me.

Thank you so much!

**Comment by ndbroadbent** [comment]
I was struggling with `Net::ReadTimeout` errors after upgrading to Google Chrome and chromedriver `75.0.3770.90`.

I tried all of the options that were mentioned, but nothing helped:

```
options.add_argument('--enable-features=NetworkService,NetworkServiceInProcess')
options.add_argument('--disable-features=VizDisplayCompositor')
options.add_argument('--disable-dev-shm-usage')
```

(It did work if I ran as the root user in my Docker container, or if I used the `--no-sandbox` flag. But those aren't good solutions.)

The solution was to upgrade from capybara `3.22.0` to `3.24.0`. (That was the only change I needed to make.) Hope that helps someone else!

**Comment by twalpole** [comment]
@ndbroadbent I'm glad to hear things are working for you, but that really is nothing between 3.22 and 3.24  - https://github.com/teamcapybara/capybara/compare/3.22.0...3.24.0 - that should have any affect on network timeouts. Are you 100% sure no other gem / chrome / chromedriver updates occurred??

**Comment by ndbroadbent** [comment]
@twalpole You're right, I downgraded `capybara` to `3.22.0`, and ran the tests in the same Docker container, and they were still passing with no `Net::ReadTimeout` errors. Then I tried to restart the Docker container, and the tests were still passing.

That's very weird, because the tests did start passing immediately after I ran `bundle update capybara`. `capybara` was the only changed gem in my Gemfile.lock, and I didn't change anything else. I had also reverted all the changes to my capybara config. Chrome and chromedriver are still on `75.0.3770.90`.

So I think this is probably just a weird Docker issue, and Chrome wasn't able start properly in the container. So if anyone else runs into that, I guess try restarting Docker first, or rebuild your images.
